server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java

```diff
@@ -192,6 +190,94 @@ private Mapping toMapping(Username rewrittenUsername, Type type) {
         throw new IllegalArgumentException("unhandled enum type");
     }
 
+    @Override
+    public List<MappingSource> getResolvedMappingSources(String user, Domain domain, EnumSet<Type> mappingTypes) throws ErrorMappingException, RecipientRewriteTableException {
+        return getMappingSources(Username.fromLocalPartWithDomain(user, domain), mappingLimit, mappingTypes);
+    }
+
+    private List<MappingSource> getMappingSources(Username username, int mappingLimit, EnumSet<Type> mappingTypes) throws TooManyMappingException, RecipientRewriteTableException {
+
+        // We have to much mappings throw ErrorMappingException to avoid
+        // infinity loop
+        if (mappingLimit == 0) {
+            throw new TooManyMappingException("554 Too many mappings to process");
+        }
+
+        Multimap<Mapping, MappingSource> mappingSources = getAllMappingSourcesForTypes(mappingTypes);
```

Comment:
This sounds expensive once you have a few million users...

Reply:
We are clearly not targeting that yet. We should not over engineer

Reply:
Well I think 'this implementation' is expensive with already just a few thousand aliases.

I would be curious to run a bench on 1k users with 1 alias each. Then 2k users with 1 alias each.